Sources/NIOCore/AsyncChannel/AsyncChannelInboundStreamChannelHandler.swift

```diff
@@ -67,10 +67,61 @@ internal final class NIOAsyncChannelInboundStreamChannelHandler<InboundIn: Senda
     @usableFromInline
     let closeRatchet: CloseRatchet
 
+    /// A type indication what kind of transformation to apply to reads.
+    @usableFromInline
+    enum Transformation {
+        /// A synchronous transformation is applied to incoming reads. This is used when bootstrapping
+        case sync((InboundIn) throws -> ProducerElement)
+        /// In the case of protocol negotiation we are applying a future based transformation where we wait for the transformation
+        /// to finish before we yield it to the source.
+        case protocolNegotiation((InboundIn) -> EventLoopFuture<ProducerElement>)
+    }
+
+    /// The transformation applied to incoming reads.
+    @usableFromInline
+    let transformation: Transformation
+
+    @inlinable
+    init(
+        eventLoop: EventLoop,
+        closeRatchet: CloseRatchet,
+        transformationClosure: @escaping (InboundIn) throws -> ProducerElement
+    ) {
+        self.eventLoop = eventLoop
+        self.closeRatchet = closeRatchet
+        self.transformation = .sync(transformationClosure)
+    }
+
     @inlinable
-    init(eventLoop: EventLoop, closeRatchet: CloseRatchet) {
+    init(
+        eventLoop: EventLoop,
+        closeRatchet: CloseRatchet,
+        protocolNegotiationClosure: @escaping (InboundIn) -> EventLoopFuture<ProducerElement>
+    ) where InboundIn == Channel {
         self.eventLoop = eventLoop
         self.closeRatchet = closeRatchet
+        self.transformation = .protocolNegotiation { channel in
+            return protocolNegotiationClosure(channel)
+                // We might be on a future from a different EL so we have to hop to the channel here.
+                .hop(to: channel.eventLoop)
+                .flatMapErrorThrowing { error in
+                    // When protocol negotiation fails the only thing we can do is
+                    // to fire the error down the pipeline and close the channel.
+                    channel.pipeline.fireErrorCaught(error)
+                    channel.close(promise: nil)
+                    throw error
+                }
+        }
+    }
+
+    @inlinable
+    init(
+        eventLoop: EventLoop,
+        closeRatchet: CloseRatchet
+    ) where InboundIn == ProducerElement {
+        self.eventLoop = eventLoop
+        self.closeRatchet = closeRatchet
+        self.transformation = .sync { $0 }
```

Comment:
I tried this initially but this doesn't work because the compiler can't infer that if the `transformation` is `none` that `InboundIn == ProducerElement`. The compiler should be smart enough here though to see that the closure is a no-op in this case and be able to optimise it away.

Reply:
There's no way the compiler makes that optimization, but it's ok, the identity function is going to be pretty cheap, and this is an implementation detail we can fix.